### PR TITLE
Add no-pare option

### PR DESCRIPTION
**Pull request by jvasile**
I've reformulated this as a no-pare option instead of an include-from-file option:

Specify review numbers in a file named FNAME, one per line, then pass
'--no-pare=FNAME --pare=100' to fix-csv.  It will pare as usual except
that if an item is on the no-pare list, it won't pare it.  This allows
the speed-up of paring but makes sure interesting items end up in the
output, which is good for debugging and testing.

### CODE REVIEW

**Comment by kfogel** [comment]
I like the idea of this.  I didn't realize the PR was here; we should re-evaluate it in light of recent code changes (like pull request #45, which is urgent and which I'm reviewing first).

**Comment by frankduncan** [comment]
I've rebased this and fixed up after the changes needed for the 2019 dataset.  @kfogel @jvasile you guys can look at this, or just merge.

The only two changes were that there was a small bug where it would fail if there was no no-pare passed in (the default NoneType needed to be an empty array to not fail the `if ... in ...` check), and not included the --pare/--no-pare options by default in wiki-refresh.

Otherwise, I've tested this with `--pare=10000`, with a file containing only 860 that was passed in as the `--no-pare` option.

**Comment by frankduncan** [comment]
I've updated this further to be --overrides rather than --no-pare as discussed here: https://github.com/OpenTechStrategies/MacFound/issues/46#issuecomment-525354589

**Comment by kfogel** [comment]
Heh.  We should either update this & merge it, or close the PR if it's obsolete.  @frankduncan, any thoughts?

(If you would need to deep-dive into this in order to collect some thoughts, and would prefer that I do that deep-dive instead, that's fine, just let me know; I'm asking you here on the off chance that your thoughts are all ready and would not require further research to express.)

**Comment by frankduncan** [comment]
This is not obsolete if we want to propagate the change to all the etl pipelines.  We can also wait for the etl pipelines to be refactored to be in one place and then this change would only be in that place.

**Comment by kfogel** [comment]
Got it.  Thanks, @frankduncan.  Let's leave this PR be for now, then.

**Comment by frankduncan** [comment]
The return of the no-pare pull request after the ETL refactor is done!  @kfogel let's review this one last time.  I'll merge it sometime this week if you don't have a chance.

File: 100Change2017/etl/compose-and-upload [hunk 1]
```diff
@@ -41,8 +41,10 @@ Command-line options:
                                   the torque wiki.  We don't automatically do that because we want to
                                   overwrite the configuration out there.
 
-  --pare N                        Reduce the number of items to 1/N rows, for quick
-                                  testing and development
+  --pare ARG                      If ARG is a number, reduce the number of items to 1/ARG.  If
```
Comment (kfogel) [inline 100Change2017/etl/compose-and-upload:44]: It might be good to be explicit about things: if you use "+" or "@" to specify which rows to include, then no *other* keys are included.  I.e., it's not just "include", it's "include only".

File: INSTALL.md [hunk 2]
```diff
@@ -166,6 +166,16 @@ $ cd <competition>/etl
 $ ./deploy ~/data
 ```
 
+Note, running `deploy` with no arguments will give you the list of args available
+or required.  In some projects, more than just the `data` directory is needed.
+
+For most scripts, both `-c` and `-p` options are available.  The former uploads
+the spreadsheet only, not uploading attachments or creating wiki pages.  For the
```
Comment (kfogel) [inline INSTALL.md:173]: Just say "The latter allows...", not "For the latter, allows..."  (I.e., might as well match the construction you use for "The former...".)

File: INSTALL.md [hunk 3]
```diff
@@ -166,6 +166,16 @@ $ cd <competition>/etl
...
+latter, allows you to include a paring option.  This option, if a number, reduces
+the set of keys to `1/NUMBER`.  If it starts with a `+`, you can add a number of
+comma separated keys to pare two, and if it starts with an `@`, then it is a file
```
Comment (kfogel) [inline INSTALL.md:176]: "to" not "two", but also you may want to reword it to be more explicit anyway.

File: etl/etl/competition.py [hunk 4]
```diff
@@ -28,7 +28,7 @@ def __init__(
         column types.  This is useful when the incoming spreadsheet was previously
         generated for torque.
 
-        Int PARE, when passed in, restricts the number of items by the factor PARE"""
+        PARE, when passed in, restricts the number of items as defined by utils.pare_pare"""
```
Comment (kfogel) [inline etl/etl/competition.py:31]: "parse_pare"

File: etl/etl/utils.py [hunk 5]
```diff
@@ -121,3 +121,27 @@ def commaize_number(number):
...
     except Exception as e:
         return number
+
+
+def parse_pare(pare_option):
+    """Parses the PARE_OPTION and returns a tuple of
+    (PARE_FACTOR, KEYS_TO_INCLUDE) where PARE_FACTOR is an int
+    by which proposals should be reduced, and KEYS_TO_INCLUDE
+    is a list of proposals to include by key.  Only one of the
+    two will ever be returned."""
```
Comment (kfogel) [inline etl/etl/utils.py:131]: No, they'll both be returned, it's just that at least one of them will be None :-).  (But seriously, it's important to be specific about that, since it's the calling discipline and in fact you do depend on the "is not None" test in the caller.)

File: etl/etl/utils.py [hunk 6]
```diff
@@ -121,3 +121,27 @@ def commaize_number(number):
...
+        pass
+    elif pare_option.startswith("@"):
+        with open(pare_option[1:]) as pare_file:
+            keys_to_include = [l.strip().split(' ')[0] for l in pare_file.readlines() if l.strip()]
+    elif pare_option.startswith("+"):
+        keys_to_include = pare_option[1:].split(",")
+    else:
+        try:
+            pare_factor = int(pare_option)
+        except Exception:
```
Comment (kfogel) [inline etl/etl/utils.py:145]: Is `except Exception` different from just `except`?  (Maybe my Python-fu is out of date?)

Comment (frankduncan) [reply etl/etl/utils.py:145]: I didn't know about that, cool!

**Comment by frankduncan** [comment]
Thanks for the review @kfogel!  I added a commit that addressed the issues, and will merge as soon as you give the thumbs up or 24 hours expires :)

File: etl/etl/utils.py [hunk 7]
```diff
@@ -128,7 +128,7 @@ def parse_pare(pare_option):
     (PARE_FACTOR, KEYS_TO_INCLUDE) where PARE_FACTOR is an int
     by which proposals should be reduced, and KEYS_TO_INCLUDE
     is a list of proposals to include by key.  Only one of the
-    two will ever be returned."""
+    two will ever be a value, forcing the other to be None."""
```
Comment (kfogel) [inline etl/etl/utils.py:131]: This still doesn't leave clear whether `(None, None)` can happen (I think it can, but the current language implies -- though doesn't quite outright state -- the opposite).  I think it's better to spend more words and lay out all three possible return states, so the caller is in no doubt.

**Comment by frankduncan** [comment]
All cleaned up.  Thanks for the review!
